src/devices/Nmea0183/Ais/AisParser.cs

```diff
@@ -0,0 +1,297 @@
+                ret.Add(rs);
+                blockNumber++;
+            }
+
+            return ret;
+        }
+
+        private bool IsValidAisSentence(RawSentence rs)
+        {
+            return rs.Valid && rs.Fields.Length == 6 && (rs.SentenceId == VdoId || rs.SentenceId == VdmId) &&
+                   rs.TalkerId == TalkerId.Ais;
+        }
+
+        private Payload? DecodePayload(string encodedPayload, int numFragments, int fragmentNumber, int messageId, int numFillBits)
+        {
+            if (numFragments == 1)
+            {
+                var decoded = _payloadDecoder.Decode(encodedPayload, numFillBits);
+                return decoded;
+            }
+
+            lock (_fragments)
+            {
+                if (fragmentNumber == 1)
+                {
+                    // Note this clears any previous message parts, which is intended (apparently the previous group with this messageId was never completed)
+                    var l = new List<string>(numFragments) { encodedPayload };
+                    _fragments[messageId] = l;
+                    return null;
+                }
+
+                if (fragmentNumber <= numFragments)
+                {
+                    if (_fragments.TryGetValue(messageId, out var existingParts) && existingParts.Count == fragmentNumber - 1)
+                    {
+                        existingParts.Add(encodedPayload);
+                    }
+                    else
+                    {
+                        // Message is incomplete or out of order -> drop it
+                        _fragments.Remove(messageId);
+                        return null;
+                    }
+                }
+
+                if (fragmentNumber == numFragments)
+                {
+                    if (_fragments.TryGetValue(messageId, out var existingParts) &&
+                        existingParts.Count == numFragments)
+                    {
+                        // The collection is complete.
+                        encodedPayload = string.Join(string.Empty, existingParts);
+                        _fragments.Remove(messageId);
+                        return _payloadDecoder.Decode(encodedPayload, numFillBits);
+                    }
+                }
+
+                return null; // More parts expected
+            }
+        }
+
+        public int ExtractChecksum(string sentence, int checksumIndex)
+        {
+            var checksum = sentence.Substring(checksumIndex + 1);
+            return Convert.ToInt32(checksum, 16);
+        }
+
+        public int CalculateChecksum(string sentence)
+        {
+            var data = sentence.Substring(1);
+
+            var checksum = 0;
+            foreach (var ch in data)
+            {
+                checksum ^= (byte)ch;
+            }
+
+            return Convert.ToInt32(checksum.ToString("X"), 16);
```

Comment:
I think I'm missing something on this line, are we converting to base 16 string and then convert it back to int?

This CalculateSum seems a bit bad from my perspective, i.e. "aa" will produce checksum 0. "bb" as well.... The answer seems to be always 1 byte, it seems it's missing some shift somewhere

Reply:
I removed this implementation, because it was a duplicate. The algorithm however was correct. This checksum is defined as the XOR of all the characters in the (ASCII) string. Not a very solid check, indeed.